Update library to work with API version 5

bachya opened the pull request:
This PR reworks the library to be compatible with MyQ API version 5. It fixes https://github.com/arraylabs/pymyq/issues/22 and fixes https://github.com/arraylabs/pymyq/issues/25. Because it is a huge rewrite (representing a completely new API), this should bump the version to 2.0.0 IMO.

In addition to this rework, this PR performs some cleanup of the library:

* Removes unnecessary `asyncio` locks
* Removes unnecessary endpoint constants
* Removes unnecessary methods to open and close a websession
* Removes unnecessary `login_request` parameter to the `_request` method
* Simplifies retry logic
* Adds `black` for nicer formatting consistent with Home Assistant (formatting, `pylintrc`, and `.flake8`)

Lastly, this PR removes a lot of the retry/rate limiting/etc. logic that was present, as that should really be the responsibility of the implementing party (Home Assistant, etc.).

bachya committed:
Update library to work with API version 5

bachya committed:
Updates

bachya committed:
More updates

bachya committed:
Docs and cleanup

bachya committed:
Cleanup

bachya committed:
Catch invalid credentials

bachya committed:
Simplified exceptions

ehendrix23 commented:
Hey Aaron,

Just to confirm as I worked on this one before as well. Number of pieces of logic in there were to reduce the actual calls to MyQ as it was determined it would start giving timeouts etc.. if too many calls would be send within a certain time. This was especially originally noticed by those with multiple garage doors as a request would be sent for each garage door.
Thus also locks to prevent multiple calls coming from Home Assistant while one is already in progress together with tracking when last actual request to MyQ was made.

Further, when opening/closing the state would still be "open" if closing the garage door, or "closed" when opening the garage door when it was being opened/closed. 
Hence logic was added to return "Opening" or "Closing" for x number of seconds before polling actual state again.

Haven't really looked at the changes but wanted to give you some background why certain pieces were in there as it was to fix issues we had when just polling MyQ each time a poll came in from HA.

Thank you for jumping so quickly on the API change!!!

bachya commented:
@ehendrix23 Thanks so much for this info; I had wondered if I was hurting something critical. I'll go back and apply your principles to this PR; perhaps you can help me continue to review once they're in?

garyak commented:
I get these a lot.

```
2019-08-07 12:09:07 WARNING (MainThread) [pymyq.api] 12:09:01.002084 Error requesting from api/v4/UserDeviceDetails/Get; retrying: 500, message='Internal Server Error'
2019-08-07 12:09:11 WARNING (MainThread) [homeassistant.helpers.entity] Update of cover.small_door is taking over 10 seconds
2019-08-07 12:09:11 WARNING (MainThread) [homeassistant.helpers.entity] Update of cover.big_door is taking over 10 seconds
2019-08-07 12:09:12 WARNING (MainThread) [pymyq.api] 12:09:01.002084 Error requesting from api/v4/UserDeviceDetails/Get; retrying: 500, message='Internal Server Error'
2019-08-07 12:09:17 WARNING (MainThread) [homeassistant.components.cover] Updating myq cover took longer than the scheduled update interval 0:00:15
2019-08-07 12:09:17 ERROR (MainThread) [pymyq.api] Getting device states failed: 12:09:01.002084 Constant timeouts while requesting data from api/v4/UserDeviceDetails/Get
```
Two doors.

bachya commented:
@garyak That's a Home Assistant log. There is non-trivial work that needs to be done for this PR to work in HASS, so if you just patched in this new code, I wouldn't expect it to work.

garyak commented:
Just an FYI. I thought it might support the earlier logic discussion. MyQ server response isn't the best.

bachya committed:
ehendrix23 suggestions

bachya commented:
@ehendrix23 Added locks and rate-limiting in place; initial tests look good. Let me know your thoughts?

ehendrix23 commented:
> @ehendrix23 Added locks and rate-limiting in place; initial tests look good. Let me know your thoughts?

Had a quick look at it and some potential observations (but let me know if they're valid or not.).

1. If I recall correctly the update of state (myqdevice.update) will be executed through HA for each cover individually. For 2 covers, HA would be executing that method twice every xxx seconds.
Now in the api.update_device_info you do throttle; but what could then happen is that we have 2 calls waiting (sleep) and then execute one after the other. Might just be better to only execute if the request if the last update has passed otherwise just ignore.

2. A loop was in there when sending requests (api.request) as sometimes things like timeouts or so would occur. This could, for example, result in garage doors not opening or closing. Hence if errors were returned we would try multiple times then with an increasing timeout for the call (if error was timeout) or waiting 5 seconds and then re-trying the request.

3. websession is still optional but does not look like if not provided one is created. That would then result in failures if none has been passed yet then trying to use it. Attriibutes we had for the web session were then also to ensure they would get cleaned up based on platform (HASSIO especially seems to have issues with this if not using the right parameters).

Everything else all seems to be based on the new API etc.. and looks good from my perspective. :-)

Is there already a PR for HA for this?

DConcord commented:
Hi @ehendrix23 and @bachya, looks like this has become critical since API v4 has is now broken. Any luck testing/validating so it can be merged? Thanks for all the good work!!

bachya commented:
Had life and work get in the way. I’ll get back in this PR tomorrow and wrap it up. 👍🏻

bachya committed:
Code review comments

bachya commented:
Well, crummy news: the API that this PR was based on no longer works exactly the same. The endpoint to get security tokens (`https://api.myqdevice.com/api/v5/Login`) returns this response:

```
{
  "code": "401.413",
  "message": "Unauthorized",
  "description": "We are no longer supporting this app. Please download our new app \"myQ Garage & Access Control\""
}
```

Digging further.

bachya commented:
It looks like Chamberlain may have implemented certificate pinning in their new iOS app (i.e., they're specifically validating the app's root certificate). This means that thus far, I've not been able to sniff traffic coming from it:

![IMG_EDFC02791F28-1](https://user-images.githubusercontent.com/47216/67171037-5fc3ce80-f372-11e9-882e-c8c582a276fc.jpeg)

bachya commented:
More info: looks like [an adventurous user may have found a viable app ID](https://github.com/arraylabs/pymyq/issues/22#issuecomment-544310368). These appear to work. Stepping forward.

bachya committed:
More code review comments

bachya commented:
@ehendrix23 Just pushed.

1. Makes sense. I've updated this PR so that if multiple device update calls execute within a 5-second window, only the first is acted upon.
2. Also makes sense; I've added in some retry logic to each API call.
3. I'm not clear on why this is necessary? Every integration I've seen in HASS requires an `aiohttp` websession to be passed in (i.e., I don't see many/any creating their own, needing special connectors, etc.).

brbeaird commented on pymyq/api.py:
```diff
@@ -1,293 +1,214 @@
 
+NON_COVER_DEVICE_FAMILIES = "gateway"
+
 BRAND_MAPPINGS = {
```
I believe these have all been rolled into a single AppID. Try just using JVM/G9Nwih5BwKgNCjLxiFUQxQijAebyyg8QUHr7JOrP+tuPb8iHfRHKwTmDzHOu

bachya replied:
bb1d5a4 should cover this.

brbeaird commented:
Yeah, the SSL pinning is a bummer. It may still be possible to load up the Android version, decompile it, remove the pinning, then debug it. That's on my to-do list of things to try this week to try and figure out how the lamp module calls look. I don't understand why sending {action_type: on} wouldn't work.

bachya committed:
Consolidate disparate app IDs into one (based on new API)

bachya committed:
Updated docs

bachya committed:
Cleanup

bachya commented:
(FYI, this should fix #25, too.)

bachya committed:
Linting

bachya committed:
Cleanup

bachya committed:
Made retry timing a tad more nuanced

bachya committed:
Simplified conditional

ehendrix23 commented:
@bachya,

Thanks for all the work. Just wanna get confirmation that his one is good and I can merge request. 
Is it then ready as well to be pushed to PIP as a new version?

bachya commented:
@ehendrix23 Confirmed: I've done a lot of testing and everything appears good to go. I think we can merge and publish this. I'll take ownership of adding it to HASS once a new PyPI package is created. That work?

ehendrix23 commented:
That works, going to see if I can create the PyPi package now. :-)

ehendrix23 commented:
PyPi package updated. Put version to 2.0.0 figuring we're switching to V5 of the MyQ API. :-)

bachya commented:
Perfect! Thank you!

jsenurquhart commented:
As an outsider looking in, I see you guys work great together. 
Thanks for all your contributions and solid work efforts!

brubakerjeff commented:
As a side, I put in a few tickets to myq support (asking about API access or ability to communicate to local hub without going to the cloud).  My understanding @bachya is having to reverse engineer it?  If so I'm thinking would be nice to have official support from the vendor **waves if someone from Chamberlain support is reading this**.

bachya commented:
>My understanding @bachya is having to reverse engineer it?

Correct: the API is reverse-engineered and it is cloud-based.

broyuken commented:
Considering they charge money to use google assistant and IFTTT, I doubt they’d want to partner with HA. They’re one of, if not the only company to have a subscription for this type of stuff and it’s pathetic. 

Edit: Thanks for the great work guys, do we have an ETA for the home assistant update?

civarson commented:
How can a normal HA (Hassio) user get this working again? Will it "just work" eventually?

bachya commented:
Update is live: https://www.home-assistant.io/blog/2019/10/10/release-100/#release-01003---october-21

difelice commented:
It's working for me. Thanks!

## Note
A new entity was created called `cover.garage_door_opener_2`, you may need to rename it (deleting old one first) if you have associated UI or automations.

bakerbug commented:
> It's working for me. Thanks!
> ## Note
> 
> A new entity was created called `cover.garage_door_opener_2`, you may need to rename it (deleting old one first) if you have associated UI or automations.

I was about to report that the update didn't work for me.  Thanks @difelice for posting that note!  I deleted my original entities, and renamed the `_2` ones, and now I'm back up an running.

blamaz commented:
It's also working for me despite my log beeing spammed with this error 

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 268, in async_update_ha_state
    await self.async_device_update()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 446, in async_device_update
    await self.async_update()
  File "/usr/src/homeassistant/homeassistant/components/myq/cover.py", line 113, in async_update
    await self._device.update()
  File "/usr/local/lib/python3.7/site-packages/pymyq/device.py", line 132, in update
    await self._api.update_device_info()
  File "/usr/local/lib/python3.7/site-packages/pymyq/api.py", line 146, in update_device_info
    for device_json in devices_resp["items"]:
TypeError: 'NoneType' object is not subscriptable

ckhyatt commented:
Just stopped working again this afternooon:
Error while setting up platform myq
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 150, in _async_setup_platform
    await asyncio.wait_for(asyncio.shield(task), SLOW_SETUP_MAX_WAIT)
  File "/usr/local/lib/python3.7/asyncio/tasks.py", line 442, in wait_for
    return fut.result()
  File "/usr/src/homeassistant/homeassistant/components/myq/cover.py", line 53, in async_setup_platform
    myq = await login(username, password, websession)
  File "/usr/local/lib/python3.7/site-packages/pymyq/api.py", line 162, in login
    await api.authenticate(username, password)
  File "/usr/local/lib/python3.7/site-packages/pymyq/api.py", line 117, in authenticate
    login_request=True,
  File "/usr/local/lib/python3.7/site-packages/pymyq/api.py", line 85, in request
    data = await resp.json(content_type=None)
  File "/usr/local/lib/python3.7/site-packages/aiohttp/client_reqrep.py", line 1040, in json
    return loads(stripped.decode(encoding))
  File "/usr/local/lib/python3.7/json/__init__.py", line 348, in loads
    return _default_decoder.decode(s)
  File "/usr/local/lib/python3.7/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/local/lib/python3.7/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

jsenurquhart commented:
Yup, looks like mines went offline randomly a few times throughout the
day.

On Tue, Oct 29, 2019 at 7:44 PM ckhyatt <notifications@github.com> wrote:

> Just stopped working again this afternooon:
> Error while setting up platform myq
> Traceback (most recent call last):
> File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py",
> line 150, in _async_setup_platform
> await asyncio.wait_for(asyncio.shield(task), SLOW_SETUP_MAX_WAIT)
> File "/usr/local/lib/python3.7/asyncio/tasks.py", line 442, in wait_for
> return fut.result()
> File "/usr/src/homeassistant/homeassistant/components/myq/cover.py", line
> 53, in async_setup_platform
> myq = await login(username, password, websession)
> File "/usr/local/lib/python3.7/site-packages/pymyq/api.py", line 162, in
> login
> await api.authenticate(username, password)
> File "/usr/local/lib/python3.7/site-packages/pymyq/api.py", line 117, in
> authenticate
> login_request=True,
> File "/usr/local/lib/python3.7/site-packages/pymyq/api.py", line 85, in
> request
> data = await resp.json(content_type=None)
> File "/usr/local/lib/python3.7/site-packages/aiohttp/client_reqrep.py",
> line 1040, in json
> return loads(stripped.decode(encoding))
> File "/usr/local/lib/python3.7/json/*init*.py", line 348, in loads
> return _default_decoder.decode(s)
> File "/usr/local/lib/python3.7/json/decoder.py", line 337, in decode
> obj, end = self.raw_decode(s, idx=_w(s, 0).end())
> File "/usr/local/lib/python3.7/json/decoder.py", line 355, in raw_decode
> raise JSONDecodeError("Expecting value", s, err.value) from None
> json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/arraylabs/pymyq/pull/23?email_source=notifications&email_token=AMMQFMN5RC5FEZWS6TGV2L3QRDDF5A5CNFSM4IJ3CVJ2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOECSOJQA#issuecomment-547677376>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AMMQFMOO4V3PQNNS5WJVAI3QRDDF5ANCNFSM4IJ3CVJQ>
> .
>

bachya commented:
Recommend everyone move their commentary from this closed PR to the newly opened issue: https://github.com/arraylabs/pymyq/issues/27

satheshshiva commented:
How to toggle the lamp states?

```
{"action_type": "on"
}
```

throws 500 internal server error
